File: src/commands/cancel.ts
import {
	ActionRowBuilder,
	ApplicationCommandType,
	ButtonBuilder,
	ButtonStyle,
	ContextMenuCommandBuilder,
	MessageActionRowComponentBuilder,
	MessageComponentInteraction,
	MessageContextMenuCommandInteraction,
} from "discord.js";

import Event from "../data/Events/event.model";
export const data = new ContextMenuCommandBuilder()
	.setName("Cancel")
	.setType(ApplicationCommandType.Message);

export async function execute(
	interaction: MessageContextMenuCommandInteraction
) {
	const message = interaction.targetMessage;
	let eventName: string;

	//* Check if event has been rolled before, if it hasn't been rolled, inform user
	if (message.reactions.cache.get("👍")?.count! === 0) {
		await interaction.reply({
			content: "This event hasn't been rolled yet",
		});
		return;
	}

	if (message.content.includes("Other")) {
		eventName = message.content.substring(
			message.content.search("-") + 2,
			message.content.search(",")
		);
	} else {
		eventName = message.content.substring(
			message.content.search("OG vs"),
			message.content.search(",")
		);
	}

	const eventToBeDeleted = await Event.findOne({ eventName: eventName }).exec();

	const event = new Event({
		eventName: eventToBeDeleted?.eventName,
		eventTime: eventToBeDeleted?.eventTime,
		eventType: eventToBeDeleted?.eventType,
		gardeners: eventToBeDeleted?.gardeners,
		hours: eventToBeDeleted?.hours,
	});

	const confirm = new ButtonBuilder()
		.setCustomId("confirm")
		.setLabel("Confirm")
		.setStyle(ButtonStyle.Danger);

	const cancel = new ButtonBuilder()
		.setCustomId("cancel")
		.setLabel("Cancel")
		.setStyle(ButtonStyle.Secondary);

	const row =
		new ActionRowBuilder<MessageActionRowComponentBuilder>().addComponents(
			confirm,
			cancel
		);

	const response = await interaction.reply({
		content: "Are you sure you want to cancel the signups for this game?",
		components: [row],
		ephemeral: true,
	});

	response
		.awaitMessageComponent({ time: 30_000 })
		.then(async (componentInteraction: MessageComponentInteraction) => {
			if (componentInteraction.customId === "confirm") {
				await Event.findByIdAndDelete(eventToBeDeleted!["_id"]);
				await componentInteraction.update({
					content: "Signups cancelled successfully",
					components: [],
				});
				await message.reactions.cache.get("787697278190223370")?.remove();
			} else {
				await componentInteraction.update({
					content: "Action cancelled",
					components: [],
				});
			}
		})
		.catch(async (err) => {
			if (event.eventName) {
				await event.save();
			}
			await interaction.editReply({
				content: "Unable to cancel sign ups, please try the command again",
				components: [],
			});
			console.error(err);
		});
}

File: lib/commands/cancel.dart
import 'package:nyxx/nyxx.dart';
import 'package:nyxx_commands/nyxx_commands.dart';
import 'package:nyxx_extensions/nyxx_extensions.dart';

import '../data/events/delete_event.dart';
import '../data/events/events.dart';

final cancel = MessageCommand(
  'Cancel this event',
  options: CommandOptions(
    defaultResponseLevel: ResponseLevel(
      hideInteraction: true,
      isDm: false,
      mention: null,
      preserveComponentMessages: false,
    ),
  ),
  id(
    'Cancel',
    (MessageContext context) async {
      final message = context.targetMessage;
      final weCooEmoji = ReactionBuilder(
        name: 'OGwecoo',
        id: Snowflake(787697278190223370),
      );

      // Check if message has already been processed
      final weCooReactions =
          await message.manager.fetchReactions(message.id, weCooEmoji);

      if (weCooReactions.isEmpty) {
        await context.respond(
          MessageBuilder(content: "This event hasn't been rolled yet"),
        );
        return;
      }

      final String eventName = message.content.substring(
        message.content.indexOf("-") + 2,
        message.content.indexOf(","),
      );

      final eventUnixTime = int.parse(
            message.content.substring(
              message.content.indexOf('<t:') + 3,
              message.content.indexOf('<t:') + 13,
            ),
          ) *
          1000;

      final eventTime =
          DateTime.fromMillisecondsSinceEpoch(eventUnixTime).toUtc();

      final confirm = await context.getConfirmation(
        MessageBuilder(
            content: 'Are you sure you want to cancel signups for this event?'),
        authorOnly: true,
        styles: {
          true: ButtonStyle.danger,
          false: ButtonStyle.secondary,
        },
        timeout: Duration(seconds: 30),
      );

      if (confirm) {
        getEventId(eventName, eventTime).flatMap(deleteEvent).match(
          (error) {
            print(error.toString());
            context.respond(
              MessageBuilder(
                content: 'Something has gone wrong, please try again',
              ),
            );
          },
          (_) async {
            Future.wait([
              context.respond(
                MessageBuilder(
                  content:
                      'Successfully cancelled signup for $eventName at <t:${eventUnixTime ~/ 1000}:F> - ${message.url}',
                ),
              ),
              message.manager.deleteReaction(
                message.id,
                weCooEmoji,
              )
            ]);
          },
        ).run();
      } else {
        await context.respond(
          MessageBuilder(
            content: 'Action cancelled',
            components: [],
          ),
        );
      }
    },
  ),
);
